File: app/src/main/java/dev/ragnarok/fenrir/domain/impl/DatabaseInteractor.java
package dev.ragnarok.fenrir.domain.impl;

import java.util.ArrayList;
import java.util.List;

import dev.ragnarok.fenrir.api.interfaces.INetworker;
import dev.ragnarok.fenrir.api.model.VKApiCity;
import dev.ragnarok.fenrir.api.model.VKApiCountry;
import dev.ragnarok.fenrir.api.model.database.ChairDto;
import dev.ragnarok.fenrir.api.model.database.FacultyDto;
import dev.ragnarok.fenrir.api.model.database.SchoolClazzDto;
import dev.ragnarok.fenrir.api.model.database.SchoolDto;
import dev.ragnarok.fenrir.api.model.database.UniversityDto;
import dev.ragnarok.fenrir.db.interfaces.IDatabaseStore;
import dev.ragnarok.fenrir.db.model.entity.CountryEntity;
import dev.ragnarok.fenrir.domain.IDatabaseInteractor;
import dev.ragnarok.fenrir.model.City;
import dev.ragnarok.fenrir.model.database.Chair;
import dev.ragnarok.fenrir.model.database.Country;
import dev.ragnarok.fenrir.model.database.Faculty;
import dev.ragnarok.fenrir.model.database.School;
import dev.ragnarok.fenrir.model.database.SchoolClazz;
import dev.ragnarok.fenrir.model.database.University;
import dev.ragnarok.fenrir.util.Utils;
import io.reactivex.rxjava3.core.Single;

public class DatabaseInteractor implements IDatabaseInteractor {

    private final IDatabaseStore cache;
    private final INetworker networker;

    public DatabaseInteractor(IDatabaseStore cache, INetworker networker) {
        this.cache = cache;
        this.networker = networker;
    }

    @Override
    public Single<List<Chair>> getChairs(int accountId, int facultyId, int count, int offset) {
        return networker.vkDefault(accountId)
                .database()
                .getChairs(facultyId, offset, count)
                .map(items -> {
                    List<ChairDto> dtos = Utils.listEmptyIfNull(items.getItems());
                    List<Chair> chairs = new ArrayList<>(dtos.size());

                    for (ChairDto dto : dtos) {
                        chairs.add(new Chair(dto.id, dto.title));
                    }

                    return chairs;
                });
    }

    @Override
    public Single<List<Country>> getCountries(int accountId, boolean ignoreCache) {
        if (ignoreCache) {
            return networker.vkDefault(accountId)
                    .database()
                    .getCountries(true, null, null, 1000)
                    .flatMap(items -> {
                        List<VKApiCountry> dtos = Utils.listEmptyIfNull(items.getItems());
                        List<CountryEntity> dbos = new ArrayList<>(dtos.size());
                        List<Country> countries = new ArrayList<>(dbos.size());

                        for (VKApiCountry dto : dtos) {
                            dbos.add(new CountryEntity().set(dto.id, dto.title));
                            countries.add(new Country(dto.id, dto.title));
                        }

                        return cache.storeCountries(accountId, dbos)
                                .andThen(Single.just(countries));
                    });
        }

        return cache.getCountries(accountId)
                .flatMap(dbos -> {
                    if (!dbos.isEmpty()) {
                        List<Country> countries = new ArrayList<>(dbos.size());
                        for (CountryEntity dbo : dbos) {
                            countries.add(new Country(dbo.getId(), dbo.getTitle()));
                        }

                        return Single.just(countries);
                    }

                    return getCountries(accountId, true);
                });
    }

    @Override
    public Single<List<City>> getCities(int accountId, int countryId, String q, boolean needAll, int count, int offset) {
        return networker.vkDefault(accountId)
                .database()
                .getCities(countryId, null, q, needAll, offset, count)
                .map(items -> {
                    List<VKApiCity> dtos = Utils.listEmptyIfNull(items.getItems());
                    List<City> cities = new ArrayList<>(dtos.size());

                    for (VKApiCity dto : dtos) {
                        cities.add(new City(dto.id, dto.title)
                                .setArea(dto.area)
                                .setImportant(dto.important)
                                .setRegion(dto.region));
                    }

                    return cities;
                });
    }

    @Override
    public Single<List<Faculty>> getFaculties(int accountId, int universityId, int count, int offset) {
        return networker.vkDefault(accountId)
                .database()
                .getFaculties(universityId, offset, count)
                .map(items -> {
                    List<FacultyDto> dtos = Utils.listEmptyIfNull(items.getItems());
                    List<Faculty> faculties = new ArrayList<>(dtos.size());

                    for (FacultyDto dto : dtos) {
                        faculties.add(new Faculty(dto.id, dto.title));
                    }

                    return faculties;
                });
    }

    @Override
    public Single<List<SchoolClazz>> getSchoolClasses(int accountId, int countryId) {
        return networker.vkDefault(accountId)
                .database()
                .getSchoolClasses(countryId)
                .map(dtos -> {
                    List<SchoolClazz> clazzes = new ArrayList<>(dtos.size());

                    for (SchoolClazzDto dto : dtos) {
                        clazzes.add(new SchoolClazz(dto.id, dto.title));
                    }

                    return clazzes;
                });
    }

    @Override
    public Single<List<School>> getSchools(int accountId, int cityId, String q, int count, int offset) {
        return networker.vkDefault(accountId)
                .database()
                .getSchools(q, cityId, offset, count)
                .map(items -> {
                    List<SchoolDto> dtos = Utils.listEmptyIfNull(items.getItems());
                    List<School> schools = new ArrayList<>(dtos.size());

                    for (SchoolDto dto : dtos) {
                        schools.add(new School(dto.id, dto.title));
                    }

                    return schools;
                });
    }

    @Override
    public Single<List<University>> getUniversities(int accountId, String filter, Integer cityId, Integer countyId, int count, int offset) {
        return networker.vkDefault(accountId)
                .database()
                .getUniversities(filter, countyId, cityId, offset, count)
                .map(items -> {
                    List<UniversityDto> dtos = Utils.listEmptyIfNull(items.getItems());
                    List<University> universities = new ArrayList<>(dtos.size());

                    for (UniversityDto dto : dtos) {
                        universities.add(new University(dto.id, dto.title));
                    }

                    return universities;
                });
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/domain/impl/DatabaseInteractor.kt
package dev.ragnarok.fenrir.domain.impl

import dev.ragnarok.fenrir.api.interfaces.INetworker
import dev.ragnarok.fenrir.db.interfaces.IDatabaseStore
import dev.ragnarok.fenrir.db.model.entity.CountryDboEntity
import dev.ragnarok.fenrir.domain.IDatabaseInteractor
import dev.ragnarok.fenrir.model.City
import dev.ragnarok.fenrir.model.database.Chair
import dev.ragnarok.fenrir.model.database.Country
import dev.ragnarok.fenrir.model.database.Faculty
import dev.ragnarok.fenrir.model.database.School
import dev.ragnarok.fenrir.model.database.SchoolClazz
import dev.ragnarok.fenrir.model.database.University
import dev.ragnarok.fenrir.util.Utils.listEmptyIfNull
import io.reactivex.rxjava3.core.Single

class DatabaseInteractor(private val cache: IDatabaseStore, private val networker: INetworker) :
    IDatabaseInteractor {
    override fun getChairs(
        accountId: Long,
        facultyId: Int,
        count: Int,
        offset: Int
    ): Single<List<Chair>> {
        return networker.vkDefault(accountId)
            .database()
            .getChairs(facultyId, offset, count)
            .map { items ->
                val dtos = listEmptyIfNull(items.items)
                val chairs: MutableList<Chair> = ArrayList(dtos.size)
                for (dto in dtos) {
                    chairs.add(Chair(dto.id, dto.title))
                }
                chairs
            }
    }

    override fun getCountries(accountId: Long, ignoreCache: Boolean): Single<List<Country>> {
        return if (ignoreCache) {
            networker.vkDefault(accountId)
                .database()
                .getCountries(true, null, null, 1000)
                .flatMap { items ->
                    val dtos = listEmptyIfNull(items.items)
                    val dbos: MutableList<CountryDboEntity> = ArrayList(dtos.size)
                    val countries: MutableList<Country> = ArrayList(dbos.size)
                    for (dto in dtos) {
                        dbos.add(CountryDboEntity().set(dto.id, dto.title))
                        countries.add(Country(dto.id, dto.title))
                    }
                    if (countries.isEmpty()) {
                        networker.vkDefault(accountId)
                            .database().getCountries(false, "RU,BY,RS", null, 100)
                            .flatMap { itemsg ->
                                val dtosg = listEmptyIfNull(itemsg.items)
                                val dbosg: MutableList<CountryDboEntity> = ArrayList(dtosg.size)
                                val countriesg: MutableList<Country> = ArrayList(dbosg.size)
                                for (dto in dtosg) {
                                    dbosg.add(CountryDboEntity().set(dto.id, dto.title))
                                    countriesg.add(Country(dto.id, dto.title))
                                }
                                cache.storeCountries(accountId, dbosg)
                                    .andThen(Single.just<List<Country>>(countriesg))
                            }
                    } else {
                        cache.storeCountries(accountId, dbos)
                            .andThen(Single.just<List<Country>>(countries))
                    }
                }
        } else cache.getCountries(accountId)
            .flatMap { dbos ->
                if (dbos.isNotEmpty()) {
                    val countries: MutableList<Country> = ArrayList(dbos.size)
                    for (dbo in dbos) {
                        countries.add(Country(dbo.id, dbo.title))
                    }
                    return@flatMap Single.just<List<Country>>(countries)
                }
                getCountries(accountId, true)
            }
    }

    override fun getCities(
        accountId: Long,
        countryId: Int,
        q: String?,
        needAll: Boolean,
        count: Int,
        offset: Int
    ): Single<List<City>> {
        return networker.vkDefault(accountId)
            .database()
            .getCities(countryId, null, q, needAll, offset, count)
            .map { items ->
                val dtos = listEmptyIfNull(items.items)
                val cities: MutableList<City> = ArrayList(dtos.size)
                for (dto in dtos) {
                    cities.add(
                        City(dto.id, dto.title)
                            .setArea(dto.area)
                            .setImportant(dto.important)
                            .setRegion(dto.region)
                    )
                }
                cities
            }
    }

    override fun getFaculties(
        accountId: Long,
        universityId: Int,
        count: Int,
        offset: Int
    ): Single<List<Faculty>> {
        return networker.vkDefault(accountId)
            .database()
            .getFaculties(universityId, offset, count)
            .map { items ->
                val dtos = listEmptyIfNull(items.items)
                val faculties: MutableList<Faculty> = ArrayList(dtos.size)
                for (dto in dtos) {
                    faculties.add(Faculty(dto.id, dto.title))
                }
                faculties
            }
    }

    override fun getSchoolClasses(accountId: Long, countryId: Int): Single<List<SchoolClazz>> {
        return networker.vkDefault(accountId)
            .database()
            .getSchoolClasses(countryId)
            .map { dtos ->
                val clazzes: MutableList<SchoolClazz> = ArrayList(dtos.size)
                for (dto in dtos) {
                    clazzes.add(SchoolClazz(dto.id, dto.title))
                }
                clazzes
            }
    }

    override fun getSchools(
        accountId: Long,
        cityId: Int,
        q: String?,
        count: Int,
        offset: Int
    ): Single<List<School>> {
        return networker.vkDefault(accountId)
            .database()
            .getSchools(q, cityId, offset, count)
            .map { items ->
                val dtos = listEmptyIfNull(items.items)
                val schools: MutableList<School> = ArrayList(dtos.size)
                for (dto in dtos) {
                    schools.add(School(dto.id, dto.title))
                }
                schools
            }
    }

    override fun getUniversities(
        accountId: Long,
        filter: String?,
        cityId: Int?,
        countyId: Int?,
        count: Int,
        offset: Int
    ): Single<List<University>> {
        return networker.vkDefault(accountId)
            .database()
            .getUniversities(filter, countyId, cityId, offset, count)
            .map { items ->
                val dtos = listEmptyIfNull(items.items)
                val universities: MutableList<University> = ArrayList(dtos.size)
                for (dto in dtos) {
                    universities.add(University(dto.id, dto.title))
                }
                universities
            }
    }
}